### PR TITLE
Replace torch einsum with opt_einsum

### PR DESCRIPTION
**Pull request by sayanghosh**
Differential Revision: D37128344

### CODE REVIEW

**Comment by sayanghosh** [comment]
Optimized einsums obtain significant trends towards an effective reduction in runtime. The most impact is from convolution and linear layers uniformly for all layer sizes. This is under 10 iterations of benchmarking. Runtime_x is from Torch einsums and runtime_y is from opt einsums (in table below).

image
image
image
image

File: opacus/grad_sample/conv.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 import torch
 import torch.nn as nn
 from opacus.utils.tensor_utils import unfold2d, unfold3d
+from opt_einsum import contract
```
Comment (karthikprasad) [inline opacus/grad_sample/conv.py:22]: `contract` not used in this file?

**Comment by sayanghosh** [comment]
Attached a notebook of the benchmarking and runtime comparison analysis here : https://colab.research.google.com/drive/1WAjSYzyKg7UisShNNwD4FMv6tIMGCzuL#scrollTo=OY0TWxFuXRwt

File: requirements.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
 numpy>=1.15
 torch>=1.8
 scipy>=1.2
+opt-einsum==3.3.0
```
Comment (karthikprasad) [inline requirements.txt:4]:
```suggestion
opt-einsum>=3.3.0
```

**Review by karthikprasad — APPROVED** [review]
Thanks. Looks good to me. Please fix the lint error before merging.

**Pull request closed without merging**